### PR TITLE
fix(prototyper): validate DBCN physical buffers by SBI ABI

### LINKED ISSUE #195: [BUG] DBCN physical address reconstruction is fixed to 32/32 bits but SBI defines it as 2×XLEN
# Summary
The fix for #192 correctly adds range validation before creating a raw slice, but it reconstructs the DBCN physical address as:

```rust
fn physical_addr(lo: usize, hi: usize) -> u64 {
    ((hi as u64) << 32) | (lo as u64)
}
```

This hard-codes the address layout as high 32 bits + low 32 bits -> u64, which does not match the SBI ABI model for physical-address arguments. In SBI, such arguments are carried as a 2×XLEN-wide quantity in two registers, where lo is the lower XLEN bits and hi is the upper XLEN bits.

As a result, the current implementation is not ABI-correct even though it works for the common case where phys_addr_hi == 0.
# Why this is BUG
The code added in #193 does two things:
- It reconstructs the address with ((hi as u64) << 32) | (lo as u64).
- It validates the resulting [start, start + len) range before from_raw_parts.

The second part is good and fixes the original hang in #192. However, the first part assumes that the two argument registers are always two 32-bit halves of a 64-bit address. That assumption is not what SBI specifies.

For SBI physical-address style parameters, the ABI meaning is:
- lo = lower XLEN bits
- hi = upper XLEN bits
So on RV32 this is a 64-bit quantity formed from two 32-bit registers, while on RV64 it is a 128-bit quantity formed from two 64-bit registers. Reconstructing it as a fixed u64 with a fixed << 32 loses that meaning.

# Impact
This is mostly hidden today because many callers pass phys_addr_hi == 0, so the incorrect reconstruction does not change the effective address in common cases.
But the current code still bakes in the wrong ABI model:
- it fixes the split at 32 bits instead of XLEN
- it fixes the full value width at 64 bits instead of 2×XLEN
That means the implementation is spec-inaccurate and can reject or mis-handle valid addresses on platforms where the upper half is meaningful.

### PR DESCRIPTION
**Pull request by Plucky923**
## Summary

This PR fixes #195  DBCN physical-buffer validation in the prototyper to follow the SBI physical-address ABI, and adds a QEMU regression test for the affected RV64 edge case.

### Prototyper

- reconstruct the DBCN physical address from `(phys_addr_lo, phys_addr_hi)` using the SBI `2 x XLEN` model instead of a hard-coded `u64` / `<< 32` split
- validate the decoded half-open range before creating raw slices
- return `invalid_address` for malformed SBI address tuples
- return `failed` for buffers that are well-formed
  but outside the platform memory window accepted by this implementation
- keep `read` and `write` on the same physical-buffer validation path
- avoid infinite retry in `fmt::Write::write_str` when the console makes no progress

### Tests / CI

- add an RV64 DBCN regression case that uses a non-zero upper address half
- fail the test if that request is incorrectly accepted
- extend the existing prototyper QEMU boot check to assert that both the non-zero upper-half write and non-zero upper-half read are rejected

## Why

The previous implementation #193  reconstructed DBCN addresses as if they were always `high 32 bits + low 32 bits -> u64`.

That does not match the SBI ABI for physical-address parameters,which are defined as `phys_addr_lo` plus `phys_addr_hi` forming a `2 x XLEN` quantity.

In the common `phys_addr_hi == 0` case this usually stayed hidden,but on RV64 it could mis-handle inputs whose upper half was non-zero.

### CODE REVIEW

File: prototyper/prototyper/src/sbi/console.rs [hunk 1]
```diff
@@ -1,33 +1,28 @@
...
-    if start > usize::MAX as u64 {
-        return false;
+fn checked_physical_addr(lo: usize, hi: usize, len: usize) -> Result<usize, SbiRet> {
+    let start = (lo as u128) | ((hi as u128) << usize::BITS);
+    let end = start
+        .checked_add(len as u128)
+        .ok_or_else(SbiRet::invalid_address)?;
+
+    if start > usize::MAX as u128 || end > usize::MAX as u128 {
+        return Err(SbiRet::failed());
```
[the remark this answers is not in the document]
Comment (Plucky923) [reply prototyper/prototyper/src/sbi/console.rs]: In SBI spec docs,
> An accessible address is one that S-mode could reasonably expect to access per its description of the platform’s physical memory layout. As an SBI implementation may further restrict the allowed range, it may return a generic SBI_ERR_FAILED (instead of SBI_ERR_INVALID_ADDRESS) when input is inaccessible with respect to its specific limits. Returning SBI_ERR_FAILED instead of SBI_ERR_INVALID_ADDRESS, in this case, is not a violation of the above specification because the SBI implementation should detect the distinct case of violating the more strict range first, making it appropriate to return the error associated with the stricter range case immediately.

I think maybe the following code is better
```rust
#[inline]
fn checked_physical_addr(lo: usize, hi: usize, len: usize) -> Result<usize, SbiRet> {
    // If the address exceeds our implementation's native `usize` capacity,
    // return SBI_ERR_FAILED as we are enforcing a stricter range limit.
    if hi != 0 {
        return Err(SbiRet::failed());
    }
    
    // Check for native usize overflow. If it overflows, it's definitively an invalid address.
    let _end = lo.checked_add(len).ok_or_else(SbiRet::invalid_address)?;

    Ok(lo)
}
```

**Review by luojia65 — APPROVED** [review]

**Review by woshiluo — APPROVED** [review]
LGTM
